### ISSUE TITLE
Split classes.xml?

### ISSUE THREAD

**Issue by chuckeles** [godotengine/godot#3941]
Currently the `doc/base/classes.xml` is 43065 lines long and has 1.22 MB on my system. A file this long makes it harder to edit, takes a lot of memory when loaded in editor and operations of type "find and highlight all words that are the same as currently selected" take quite some time.

Wouldn't it be better to split this file into multiple? For example one per class? module?

**Comment by est31** [comment]
I don't have problems with kate here (never had any problems with large files at all, except for the line limit issue with kate), but its better if we support as much editors as possible.

**Comment by akien-mga** [comment]
I wouldn't split it in one file per class, that would make it much harder to edit for me. Many classes reuse similar functions (e.g. AudioServer and the various *SamplePlayer) for which I often had to ensure that the virtual methods of the one are coherent with the virtual methods of the other.

I guess it could be split in "Core", "Nodes", "Resoures" or something.

**Comment by chuckeles** [comment]
I use atom and some hightlighting features were a little laggy. Spliting in core, nodes, resources would be perfect.

**Comment by bojidar-bg** [comment]
@chuckeles ~ atom is rather slow for such large files, maybe use kate/gedit/notepad++?
@akien-mga ~ We can always make a godot "game" that would allow for editing of classes.xml :smile:

**Comment by akien-mga** [comment]
>  We can always make a godot "game" that would allow for editing of classes.xml :smile:

Someone had actually started doing this, but I don't really see the point. It just makes it harder not to have a true text editor for this :)

**Comment by trollworkout** [comment]
This may be quite extreme change but what about making the whole file less verbose  using a tscn like INI style instead of xml?

```
[class  name="@GDScript" category="Core"]
brief_description = "Built-in GDScript functions."
description="This contains the list of built-in gdscript functions. Mostly math functions and other utilities. Everything else is expanded by objects."

[method name="fmod" class="@GDScript"]
argument_name = {"x", "y"}
argument_type = { float, float }
return = float
description = "Module (remainder of x/y)."

[constant name="PI" value="3.141593" class="@GDScript"]
description = "Constant that represents how many times the diameter of a circumference fits around it's perimeter."
```

**Comment by bojidar-bg** [comment]
@trollworkout ~ sounds like a good idea, since may people hate XML, while we already use INI for many things. I would just propose a change:

Instead of:
````
[method name="fmod" class="@GDScript"]
argument_name = {"x", "y"}
argument_type = { float, float }
````
To do:
````
[method name="fmod" class="@GDScript"]
arguments/count = 0
arguments/0 = { "x", float }
arguments/1 = { "y", float }
````
Or maybe (if order is preserved):
````
[method name="fmod" class="@GDScript"]
arguments/x = "float"
arguments/y = "float"
````

**Comment by akien-mga** [comment]
I don't think using an INI format would be good for that. It has no indentation, so it would be quite hard to see where a class definition starts and where it ends. So we'd have to split all classes in their own files, and that's IMO much harder to maintain (and parse).

**Comment by bojidar-bg** [comment]
@akien-mga ~ Maybe JSON? Or IINI (Indented INI) :smile:? Or just, plain old YAML/CSON...

**Comment by akien-mga** [comment]
Or XML? :)

**Comment by bojidar-bg** [comment]
@akien-mga ~ Seems like that's the best format for now :(

**Comment by trollworkout** [comment]
@bojidar-bg  looking again at tscn I would say this may be even closer
```
[method name="fmod" class="@GDScript"]
arguments = { float : "x", float : "y" }
```

**Comment by bojidar-bg** [comment]
@trollworkout ~ better flip them around, since "float" is a very bad name for a Dictionary key..

**Comment by trollworkout** [comment]
@bojidar-bg  You are right. Is much easier to map to a dictionary if they are flipped.  I happen to like the first one arguments = { "x" : float, "y" : float } seems to me most logical. I find xml quite hard to read.

Other choices would be json or cson or yaml all of which seem to be a bit less verbose than xml . That plus split the classes into it's own file I guess that could do it too although would be a bit harder to track.

**Comment by akien-mga** [comment]
Fixed by 4f929a0f.
